core/schema/internalsdk.graphqls

```diff
@@ -0,0 +1,138 @@
+extend type Query {
+  """
+  Create a new function from the provided definition.
+  """
+  newFunction(def: FunctionDef!): Function!
+
+  """
+  The FunctionCall context that the SDK caller is currently executing in.
+  If the caller is not currently executing in a function, this will return
+  an error.
+  """
+  currentFunctionCall: FunctionCall!
+}
+
+type FunctionCall {
+  "The name of the function being called."
+  name: String!
+
+  """
+  The name of the parent object of the function being called.
+  If the function is "top-level" to the module, this is the name of the module.
+  """
+  parentName: String!
+
+  """
+  The value of the parent object of the function being called.
+  If the function is "top-level" to the module, this is always an empty object.
+  """
+  parent: JSON!
+
+  """
+  The argument values the function is being invoked with.
+  """
+  inputArgs: [FunctionCallArgValue!]
+
+  """
+  Set the return value of the function call to the provided value.
+  The value should be a string of the JSON serialization of the return value.
+  """
+  returnValue(value: JSON!): Void
```

Comment:
This in particular is another verb that we could consider putting under `Mutation` instead

Reply:
@vito Just saw this, yeah so the engine (server-side) knows what the return type for a given function call is going to be because the module has to tell the engine all of the type defs for all of the function args/returns before anything gets called. So if the return is id-able, it knows that ahead of time and knows to decode it. [Relevant code here](https://github.com/dagger/dagger/blob/f3edc96ab38dc308b45cc3e5404b32fb39e329b6/core/schema/module.go#L868-L876).

ID-able types in particular are json serialized to their id string, all other types are just standard JSON serialization. In the previous project/environment iterations of all this, that rule was sort of implicit and scattered around the code. I slightly formalized it a bit more now by updating the Go SDK client codegen to have explicit MarshalJSON and UnmarshalJSON methods on id-able objects that enforce all this.

---

More thoughts on this general topic: this is highly related to what we were talking about the other today in terms of whether we should make all objects ID-able or continue to reserve that for only our special core API types and use separate `input` types as needed.

After implementing the API in `internalsdk.graphqls` I'm more inclined than ever to go with the "everything-is-id-able" approach. On the surface it felt more complicated initially, but in practice trying to use input types has been kind of a pain since they seem to easily result in circular type definitions in our codegen SDKs, which has secondary/tertiary effects. Forcing the input representation of a type to be a scalar string helps avoids that, in addition to the other possible benefits around having a sort of "unified ID" system you've been talking about.

I don't think this would change that much about this `returnValue` API currently; we still need to explicitly ask for the type defs of all functions in the module ahead of time and use that knowledge to determine whether to de-ser an ID here or not. But we'll expand the scope of that rule to be "all objects" rather than just "core api objects". Basically, [relax the code here](https://github.com/dagger/dagger/blob/f3edc96ab38dc308b45cc3e5404b32fb39e329b6/core/schema/module.go#L830-L837).